Google.PowerShell/Common/AuthenticateWithSdkCredentialsExecutor.cs

```diff
@@ -0,0 +1,159 @@
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+using Google.Apis.Auth.OAuth2;
+using Google.Apis.Auth.OAuth2.Flows;
+using Google.Apis.Http;
+using System;
+using System.Net;
+using System.Net.Http;
+using System.Threading;
+using System.Threading.Tasks;
+
+namespace Google.PowerShell.Common
+{
+    /// <summary>
+    /// OAuth 2.0 credential for accessing protected resources using an access token.
+    /// This class will get the access token from "gcloud auth print-access-token".
+    /// </summary>
+    public class AuthenticateWithSdkCredentialsExecutor : ICredential, IHttpExecuteInterceptor, IHttpUnsuccessfulResponseHandler
+    {
+        /// <summary>
+        /// This is a cache of the last call to "gcloud auth print-access-token".
+        /// If the current user hasn't changed, and the token hasn't expired, we reuse it
+        /// rather than launching gcloud again.
+        /// </summary>
+        private static ActiveUserToken s_token;
+        private static readonly SemaphoreSlim s_tokenLock = new SemaphoreSlim(1);
+
+        /// <summary>
+        /// Returns the authorization code flow. 
+        /// This is used to revoke token (in RevokeTokenAsync)
+        /// and intercept request with the access token (in InterceptAsync).
+        /// </summary>
+        public IAuthorizationCodeFlow Flow { get; } = new GoogleAuthorizationCodeFlow(
+                new GoogleAuthorizationCodeFlow.Initializer()
+                {
+                    ClientSecrets = new ClientSecrets()
+                    {
+                        ClientId = "clientId",
+                        ClientSecret = "clientSecrets"
+                    }
+                });
+
+        #region IHttpExecuteInterceptor
+
+        /// <summary>
+        /// Default implementation is to try to refresh the access token if there is no access token or if we are 1 
+        /// minute away from expiration. If token server is unavailable, it will try to use the access token even if 
+        /// has expired. If successful, it will call <see cref="IAccessMethod.Intercept"/>.
+        /// </summary>
+        public async Task InterceptAsync(HttpRequestMessage request, CancellationToken taskCancellationToken)
+        {
+            Task<string> getAccessTokenTask = GetAccessTokenForRequestAsync(request.RequestUri.ToString(), taskCancellationToken);
+            string accessToken = await getAccessTokenTask.ConfigureAwait(false);
+            Flow.AccessMethod.Intercept(request, accessToken);
+        }
+
+        #endregion
+
+        #region IHttpUnsuccessfulResponseHandler
+
+        public async Task<bool> HandleResponseAsync(HandleUnsuccessfulResponseArgs args)
+        {
+            if (args.Response.StatusCode == HttpStatusCode.Unauthorized)
+            {
+                return await RefreshTokenAsync(args.CancellationToken).ConfigureAwait(false);
```

Comment:
Could you add the doc comment from `IHttpUnsuccessfulResponseHandler.HandleResponseAsync` here? I don't know the purpose of returning the result of `RefreshTokenAsync` here, and so I'm inclined to assume we are doing the wrong thing.